jsonrpc2.go

```diff
@@ -675,28 +695,7 @@ func (c *Conn) readMessages(ctx context.Context) {
 			}
 		}
 	}
-
-	c.sending.Lock()
-	c.mu.Lock()
-	c.shutdown = true
-	closing := c.closing
-	if err == io.EOF {
-		if closing {
-			err = ErrClosed
```

Comment:
Excellent observation! I definitely overlooked this before you drew my attention to it. As it turns out, from what I can see, we don't actually need to send `ErrClosed` through the channel because the [channel listener infers `ErrClosed` when the channel is being closed](https://github.com/sourcegraph/jsonrpc2/blob/master/jsonrpc2.go#L552-L555). Based on that, I'm coming to the conclusion that having the following two lines...

```go
call.done <- ErrClosed
close(call.done)
```
...is redundant and we can reduce the code to just `close(call.done)` (see 85075f0782fc4062813f12fc5d41a787c97a44ab). A follow-on question that may arise from my conclusion is "_what about errors other than `ErrClosed`? Now those don't get sent to `c.pending`_". It seems to me that if a connection is being closed, the cause of that closure is not the concern of a pending request. If a pending request is being cancelled due to the connection being closed, then all it cares about is `ErrClosed`.

Now, this investigation made me pay more attention to `TestConn_Close: waiting for response`, which is the test that I believe is responsible for verifying the behavior discussed above. That made me discover an issue with the test, which I discuss [here](https://github.com/sourcegraph/jsonrpc2/pull/64/commits/85075f0782fc4062813f12fc5d41a787c97a44ab#r1099163167).